tensorflow/core/kernels/dml_tensor_scatter_ops.cc

```diff
@@ -0,0 +1,385 @@
+    scatter_nd_desc.IndicesDimensionCount = indices_shape.dims();
+
+    DML_OPERATOR_DESC op_desc = {DML_OPERATOR_SCATTER_ND, &scatter_nd_desc};
+    Initialize(ctx, std::move(tensors), op_desc);
+  }
+};
+
+template <typename Index, typename BinaryOp>
+class DmlTensorScatterBinaryKernel : public DmlKernel {
+ public:
+  using InitHelper = TensorScatterInitHelper<Index>;
+
+  DmlTensorScatterBinaryKernel(DmlKernelConstruction* ctx,
+                               const InitHelper* init_helper) {
+    const TensorShape& in_out_shape = ctx->GetInputTensorShape(0);
+    const TensorShape& indices_shape = ctx->GetInputTensorShape(1);
+    const TensorShape& updates_shape = ctx->GetInputTensorShape(2);
+
+    DmlTensorInfo in_out_tensor;
+    in_out_tensor.desc = DmlTensorDesc::Create(ctx->GetInputDataType(0),
+                                               in_out_shape, in_out_shape);
+
+    DmlTensorInfo indices_tensor;
+    indices_tensor.desc = DmlTensorDesc::Create(ctx->GetInputDataType(1),
+                                                indices_shape, indices_shape);
+
+    DmlTensorInfo updates_tensor;
+    updates_tensor.desc = DmlTensorDesc::Create(ctx->GetInputDataType(2),
+                                                updates_shape, updates_shape);
+
+    DmlTensorInfo empty_tensor;
+    empty_tensor.desc = DmlTensorDesc::Create(ctx->GetInputDataType(0),
+                                              in_out_shape, TensorShape({1}));
+
+    DmlKernelTensors tensors;
+    tensors.inputs = {in_out_tensor, indices_tensor, updates_tensor,
+                      empty_tensor};
+    tensors.outputs = {in_out_tensor};
+
+    auto dml_dtype = GetDmlDataTypeFromTfDataType(ctx->GetOutputDataType(0));
+    constexpr uint32_t in_dim_count = 1;
+    constexpr uint32_t in_size = 1;
+    constexpr uint32_t in_stride = 1;
+    empty_buffer_size_ =
+        DMLCalcBufferTensorSize(dml_dtype, in_dim_count, &in_size, &in_stride);
+
+    auto inputs = GetDmlTensorDescs(tensors.inputs);
+    auto scope = dml::Scope(ctx->GetDmlDevice());
+    auto input = dml::InputTensor(scope, 0, inputs[0]);
+    auto indices = dml::InputTensor(scope, 1, inputs[1]);
+    auto updates = dml::InputTensor(scope, 2, inputs[2]);
+    auto empty_input = dml::InputTensor(scope, 3, inputs[3]);
+
+    // First, perform the scatter on an empty tensor
+    auto result = dml::ScatterND(empty_input, indices, updates,
+                                 in_out_shape.dims(), indices_shape.dims());
+
+    // Then, perform the binary operation on the scattered tensor and the
+    // original input tensor
+    result = BinaryOp()(input, result);
+
+    Microsoft::WRL::ComPtr<IDMLCompiledOperator> compiled_op =
+        scope.Compile(DML_EXECUTION_FLAG_NONE, {result});
+
+    Initialize(ctx, std::move(tensors), compiled_op.Get());
+  }
+
+  StatusOr<DmlGpuEvent> Compute(DmlKernelContext* ctx) const override {
+    D3D12BufferRegion input_buffer =
```

Comment:
Can we eliminate all this custom logic by using FillValueConstant to generate the initial empty scalar?

Reply:
Sure we can. My thought process was that zeroing the buffer would be faster than adding another dispatch.

Reply:
Maybe we should keep a pre-zero'd tensor attached to the device for situations such as these - I think we have several other instances where we do stuff like this. It's kinda similar to how CPUs have a zero register. Either way it's not a big deal, so I'll leave the decision up to you.